### ISSUE TITLE
Empty stream not permitted

### ISSUE THREAD

**Issue by aaliddell** [vmagamedov/grpclib#69]
Consider a service defined as the following:

```
service blah {
     rpc ListSomeObjects(SomeRequest) returns (stream SomeObject) {}
}
```

This is a common usage pattern, such as returning a directory listing and is part of the interoperability tests from core grpc: https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md#empty_stream. If no objects exist to return, the stream should return no messages, but also no error. However, with grpclib, if the stream sends no messages and ends, the library will force send a `UNKNOWN` error:

https://github.com/vmagamedov/grpclib/blob/66de865a2c1cf5d9cc147cb5761ad2d0f07f7f19/grpclib/server.py#L239-L241

Manually trying to send the trailers as `OK` with `send_trailing_metadata` also fails, as the same check exists here:

https://github.com/vmagamedov/grpclib/blob/66de865a2c1cf5d9cc147cb5761ad2d0f07f7f19/grpclib/server.py#L179-L181

To fix, the library should only automatically send `UNKNOWN` if no messages have been sent _and_ the response type is not streaming. Likewise for the check in `send_trailing_metadata`, which should allow sending OK for empty streaming responses.

There may also be similar changes needed in the client side, but I've not checked.

**Comment by vmagamedov** [comment]
Fixed and tested this behavior on the server-side and on the client-side.

**Comment by vmagamedov** [comment]
Also found an issue with trailers-only responses: #70. Maybe it should be possible to send empty response as trailers-only response for streaming methods.

**Comment by aaliddell** [comment]
Ok, thanks for the quick fix.

Perhaps a long-term goal might be to implement a test suite that runs a set of compatibility tests with 'core' GRPC as the opposite end client/server to check for differences in behaviour?

**Comment by vmagamedov** [comment]
Sure, it would be great to implement https://github.com/grpc/grpc/blob/master/doc/interop-test-descriptions.md, haven't seen this document before.

**Comment by popravich** [comment]
Just tested version from master and now it raises:
```python
GRPCError: (<Status.UNKNOWN: 2>, 'Missing content-type header')
```

**Comment by vmagamedov** [comment]
@popravich please elaborate, what you try to do and what's wrong? Server-side or client-side? You can also try `trailers-only` branch, where I'm trying to fix everything related to this issue.

**Comment by popravich** [comment]
I'm testing the following code:
```python
# client call
async def call_some_method():
    async with some_stub.List.open() as stream:
        await stream.send_request(SomeRequest())
        async for msg in stream:
            pass
# the server handler looks like this
async def List(self, stream):
    req = await stream.recv_message()
    async for obj in some_empty_iterator():
        await stream.send_message(SomeReply(data=obj))
```
And getting the following error
```python
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/lib/python3.7/dist-packages/grpclib/stream.py:40: in __anext__
    message = await self.recv_message()
/usr/local/lib/python3.7/dist-packages/grpclib/client.py:341: in recv_message
    await self.recv_initial_metadata()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <grpclib.client.Stream object at 0x7feef7d992e8>

    async def recv_initial_metadata(self):
        """Coroutine to wait for headers with initial metadata from the server.
    
        .. note:: This coroutine will be called implicitly during first
            :py:meth:`recv_message` coroutine call, if not called before
            explicitly.
    
        May raise :py:class:`~grpclib.exceptions.GRPCError` if server returned
        non-:py:attr:`Status.OK <grpclib.const.Status.OK>` in trailers-only
        response.
    
        When this coroutine finishes, you can access received initial metadata
        by using :py:attr:`initial_metadata` attribute.
        """
        if not self._send_request_done:
            raise ProtocolError('Request was not sent yet')
    
        if self._recv_initial_metadata_done:
            raise ProtocolError('Initial metadata was already received')
    
        try:
            with self._wrapper:
                headers = await self._stream.recv_headers()
                self._recv_initial_metadata_done = True
    
                metadata = decode_metadata(headers)
                metadata, = await self._dispatch.recv_initial_metadata(metadata)
                self.initial_metadata = metadata
    
                headers_map = dict(headers)
                self._raise_for_status(headers_map)
                self._raise_for_grpc_status(headers_map, optional=True)
    
                content_type = headers_map.get('content-type')
                if content_type is None:
                    raise GRPCError(Status.UNKNOWN,
>                                   'Missing content-type header')
E                                   grpclib.exceptions.GRPCError: (<Status.UNKNOWN: 2>, 'Missing content-type header')

/usr/local/lib/python3.7/dist-packages/grpclib/client.py:288: GRPCError
```

**Comment by vmagamedov** [comment]
There is a bug in a trailers-only response from the server-side, it was fixed in 8d5cc07dcfb7fb5b82c97fd50f53c904bbcf6d45, going to issue next release candidate tomorrow.

**Comment by vmagamedov** [comment]
Fixed and issued `0.3.0rc2` release candidate
